pkg/cmd/grafana-example-apiserver/server.go

```diff
@@ -0,0 +1,128 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"net"
+
+	exampleAPI "github.com/grafana/grafana/pkg/registry/apis/example"
+	grafanaAPIServer "github.com/grafana/grafana/pkg/services/grafana-apiserver"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	genericapiserver "k8s.io/apiserver/pkg/server"
+	"k8s.io/apiserver/pkg/server/options"
+	netutils "k8s.io/utils/net"
+)
+
+const defaultEtcdPathPrefix = "/registry/example.grafana.app"
+
+var (
+	Scheme = runtime.NewScheme()
+	Codecs = serializer.NewCodecFactory(Scheme)
+
+	unversionedVersion = schema.GroupVersion{Group: "", Version: "v1"}
+	unversionedTypes   = []runtime.Object{
+		&metav1.Status{},
+		&metav1.WatchEvent{},
+		&metav1.APIVersions{},
+		&metav1.APIGroupList{},
+		&metav1.APIGroup{},
+		&metav1.APIResourceList{},
+	}
+)
+
+func init() {
+	// we need to add the options to empty v1
+	metav1.AddToGroupVersion(Scheme, schema.GroupVersion{Group: "", Version: "v1"})
+	Scheme.AddUnversionedTypes(unversionedVersion, unversionedTypes...)
+}
+
+// ExampleServerOptions contains state for master/api server
+type ExampleServerOptions struct {
+	RecommendedOptions *options.RecommendedOptions
+	Builders           []grafanaAPIServer.APIGroupBuilder
+	AlternateDNS       []string
+
+	StdOut io.Writer
+	StdErr io.Writer
+}
+
+func NewExampleServerOptions(out, errOut io.Writer) (*ExampleServerOptions, error) {
+	builder := &exampleAPI.TestingAPIBuilder{}
+
+	// Install schema
+	if err := builder.InstallSchema(Scheme); err != nil {
+		return nil, err
+	}
+
+	return &ExampleServerOptions{
+		Builders: []grafanaAPIServer.APIGroupBuilder{builder},
+		RecommendedOptions: options.NewRecommendedOptions(
+			defaultEtcdPathPrefix,
+			Codecs.LegacyCodec(builder.GetGroupVersion()),
+		),
+		StdOut: out,
+		StdErr: errOut,
+	}, nil
+}
+
+func (o ExampleServerOptions) Config() (*genericapiserver.RecommendedConfig, error) {
+	if err := o.RecommendedOptions.SecureServing.MaybeDefaultWithSelfSignedCerts("localhost", o.AlternateDNS, []net.IP{netutils.ParseIPSloppy("127.0.0.1")}); err != nil {
+		return nil, fmt.Errorf("error creating self-signed certificates: %v", err)
+	}
+
+	o.RecommendedOptions.Authentication.RemoteKubeConfigFileOptional = true
+	o.RecommendedOptions.Authorization.RemoteKubeConfigFileOptional = true
+
+	o.RecommendedOptions.SecureServing.SecureServingOptions.BindPort = 8443
+
+	o.RecommendedOptions.Admission = nil
+	o.RecommendedOptions.CoreAPI = nil
```

Comment:
do we need to override any CLI options at this point?

Reply:
When deploying it on GKE / us-central-0 dev, I could see the need to pass a kubeconfig for the root cluster to authenticate with. We usually provide the same one for authz and authn file arguments. Possibly one more cert argument for authenticating to the aggregated server directly for troubleshooting.

Reply:
All these:

```
--etcd-servers=http://127.0.0.1:2379 --authentication-kubeconfig ~/.kube/config --authorization-kubeconfig ~/.kube/config --kubeconfig ~/.kube/config
```